docs/docs/typelevel.md

```diff
@@ -0,0 +1,532 @@
+# Functional Typelevel Programming
+
+This note is an exploration how we might combine staging with
+typelevel metaprogramming in Dotty.
+
+### State of the Art
+
+Currently typelevel programming uses implicits as the basic mechanism,
+resulting in a programming style much like Prolog. Amazing feats have
+been achieved using this scheme, but things are certainly far from
+ideal. In particular:
+
+ - The logic programming style requires a shift of mindset compared to the
+   usual functional programming style in Scala.
+ - The ways to control implicit search are underdeveloped,
+   leading to complicated schemes, requiring rule prioritization or other hacks.
+ - Because of their conceptual complexity the resulting typelevel programs are often
+   fragile.
+ - Error diagnostics are usually very bad, making programming with implicits somewhat
+   of a black art. Dotty has greatly improved error dignostics for recursive implicits,
+   but the fundamental problem remains.
+
+In the following we develop a system and methodology that lets us do
+typelevel programming in a functional style. The system is based on
+Dotty's symmetric meta programming system, adding some fairly modest
+language extensions.
+
+### Prequel: Conditions in Staged Code
+
+As a perparatory step, we allow if-then-else in spliced code. Given an
+if-then-else expression, the macro-interpreter interprets its
+condition (which must be a compile-time constant) and chooses one of
+the two branches depending on the condition's value. Consider the
+following example mapping an integer value `n` to a peano number
+represented as an instance of type `Nat`.
+
+    trait Nat
+
+    case object Z extends Nat
+    case class S[N <: Nat] extends Nat
```

Comment:
I’m reading and commenting as I go along. This is confusing. It is similar but not identical to “singleton types” à la Haskell (from Tim Sheard’s Ωmega), but unlike those it seems to mix type-level and value level. For numbers this seems to work, but how do you encode correctly the type `Eq[A, B]` so that it’s only inhabited if `A` and `B` are equal? It’s OK to use Peano numbers as examples, but I’d be wary of using them as evaluation.

With singletons you have both type-level numbers:
```scala
trait TNat
case object TZ extends TNat
case class TS(n: TNat) extends TNat
```

and value-level numbers, indexed by type-level numbers, so that `Nat[N]` has *one* inhabitant which is isomorphic to `N`, and so that you can operate on this number at both value and type-level since you have the number at both levels.

```
trait Nat[N <: TNat]
case object Z extends Nat[TZ]
case object S[TN](n: Nat[TN]) extends Nat[TS[TN]]
```

Reply:
There's one bit of information I glossed over here: In the definition of `case object Z` I assumed it defines a type as well as an object. I.e. that we expand to

     lazy val Z: Z = new Z
     class Z

Current Scala does not do this. We could of course always use the more explicit definition, in which case type `Nat` would become:

    trait Nat
    class Z extends Nat
    val Z = new Z
    class S[N <: Nat] extends Nat
    def S[N <: Nat] = new S[N]

So that's what I meant in painful detail when I wrote the definitions of `Nat`. The other types are analogous.

Reply:
I was assuming that `Z.type` exists; but `S[N]` surprised me because it is partly static and partly dynamic. So for instance I can’t convert `Nat` to a normal number at runtime. Anyway, I’ll try to figure out concretely what can be encoded in this style.

Reply:
> So for instance I can’t convert Nat to a normal number at runtime. 

Indeed. Type arguments are erased, so there is not enough information available at runtime.